examples/mortality.py

```diff
@@ -0,0 +1,268 @@
+
+
+def create_lookup(s1, s2):
+    """
+    Create a map between s1 indices and unique s2 indices
+    """
+    lookup = np.column_stack([s1, s2])
+    lookup = np.unique(lookup, axis=0)
+    lookup = lookup[lookup[:, 1].argsort()]
+    return lookup[:, 0]
+
+
+reparam_config = {
+    k: LocScaleReparam(0)
+    for k in [
+        "alpha_s1",
+        "alpha_s2",
+        "alpha_age_drift",
+        "beta_age_drift",
+        "pi_drift",
+    ]
+}
+
+
+@numpyro.handlers.reparam(config=reparam_config)
+def model(age, space, time, lookup, population, deaths=None):
+    N_s1 = len(np.unique(lookup))
+    N_s2 = len(np.unique(space))
+    N_age = len(np.unique(age))
+    N_t = len(np.unique(time))
+    N = len(population)
+
+    # plates
+    age_plate = numpyro.plate("age_groups", N_age, dim=-3)
+    space_plate = numpyro.plate("space", N_s2, dim=-2)
+    year_plate = numpyro.plate("year", N_t - 1, dim=-1)
+
+    # hyperparameters
+    sigma_alpha_s1 = numpyro.sample("sigma_alpha_s1", dist.HalfNormal(1.0))
+    sigma_alpha_s2 = numpyro.sample("sigma_alpha_s2", dist.HalfNormal(1.0))
+    sigma_alpha_age = numpyro.sample("sigma_alpha_age", dist.HalfNormal(1.0))
+    sigma_beta_age = numpyro.sample("sigma_beta_age", dist.HalfNormal(1.0))
+    sigma_pi = numpyro.sample("sigma_pi", dist.HalfNormal(1.0))
+
+    # spatial hierarchy
+    with numpyro.plate("s1", N_s1, dim=-2):
+        alpha_s1 = numpyro.sample("alpha_s1", dist.Normal(0, sigma_alpha_s1))
+    with space_plate:
+        alpha_s2 = numpyro.sample(
+            "alpha_s2", dist.Normal(alpha_s1[lookup], sigma_alpha_s2)
+        )
+
+    # age
+    with age_plate:
+        alpha_age_drift_scale = jnp.pad(
+            jnp.broadcast_to(sigma_alpha_age, N_age - 1),
+            (1, 0),
+            constant_values=10.0,  # pad so first term is alpha0, prior N(0, 10)
+        )[:, jnp.newaxis, jnp.newaxis]
+        alpha_age_drift = numpyro.sample(
+            "alpha_age_drift", dist.Normal(0, alpha_age_drift_scale)
+        )
+        alpha_age = jnp.cumsum(alpha_age_drift, -3)
+
+        beta_age_drift_scale = jnp.pad(
+            jnp.broadcast_to(sigma_beta_age, N_age - 1), (1, 0), constant_values=10.0
+        )[:, jnp.newaxis, jnp.newaxis]
+        beta_age_drift = numpyro.sample(
+            "beta_age_drift", dist.Normal(0, beta_age_drift_scale)
+        )
+        beta_age = jnp.cumsum(beta_age_drift, -3)
+        beta_age_cum = jnp.outer(beta_age, jnp.arange(N_t))[:, jnp.newaxis, :]
+
+    # random walk over time
+    with year_plate:
+        pi_drift = numpyro.sample("pi_drift", dist.Normal(0, sigma_pi))
+        pi = jnp.pad(jnp.cumsum(pi_drift, -1), (1, 0))
+
+    # likelihood
+    latent_rate = alpha_age + beta_age_cum + alpha_s2 + pi
+    with numpyro.plate("N", N):
+        mu_logit = latent_rate[space, age, time]
+        mu = numpyro.deterministic("mu", expit(mu_logit))
+        numpyro.sample("deaths", dist.Binomial(population, mu), obs=deaths)
```

Comment:
There are two things that helped (no divergences but slow :( ):
+ a bug in the model: `mu_logit = latent_rate[age, space, time]`
+ a bug in Binomial??: `numpyro.sample("deaths", dist.Binomial(population, logits=mu_logit), obs=deaths)`

Reply:
@theorashid The second bug is fixed in #1387. In the meantime, you can use `dist.Binomial(population, logits=mu_logit)` (recommended) instead. Hope that those changes fix your issue.

Reply:
That's great, thanks. The numypro model takes 8 mins on my laptop. The pymc model takes 14 mins using the same numpyro sampler– but I'll post on their forum now in case the devs there have any speedups.

For some reason, they are producing different results. I'm not too worried as it isn't real data, but it is weird that they are converging to different things. `alpha0` in [pymc-mod.ipynb](https://github.com/theorashid/mortality-statsmodel/tree/pymc/Models/parametric) and `alpha_age_drift_decentered[0,0,0]` in [numpyro-mod.ipynb](https://github.com/theorashid/mortality-statsmodel/tree/pymc/Models/parametric) should be the same, unless the recentring means they aren't actually comparable.

Reply:
@theorashid Probably `alpha_s1` and `alpha_s2` are not reparameterized there?

Reply:
I did the reparametrisation myself as
```
alpha_age_innovations = pm.Normal("alpha_age_innovations", sigma=1.0, dims="age_minus_origin")
alpha_age_raw_rw = aet.cumsum(aet.concatenate([aet.zeros(shape=(1,)), alpha_age_innovations], axis=-1), axis=-1)
alpha_age = pm.Deterministic("alpha_age", alpha0 + alpha_age_raw_rw * sigma_alpha_age, dims="age")
```

So I suppose it depends whether pymc's `alpha_age[0]` and numpyro's `alpha_age_drift_decentered[0,0,0]` are equivalent, or whether the latter is actually pymc's `alpha_age_raw_rw[0]` or `alpha_age_innovations[0]`.

Reply:
> Would that make a difference to the values?

We have the same model, but using reparameterization might help the sampler.

Reply:
I'll run it, but I don't think normal(0, sd) -> sd * normal(0, 1) make much difference. One is just a scaled geometry of the other (I think). I agree normal(mean, sd) -> mean + normal will make a difference. But I'll run it and see what happens.

Reply:
I've dug further into this. The parameter estimates are roughly the same **except** for `alpha0`/`beta0` in pymc and their numpyro equivalents `alpha_age_drift_decentered[0,0,0]`/`beta_age_drift_decentered[0,0,0]`. I'm not quite sure why but I don't think it's important. So I'll update the model in this fork and move on